helix-core/src/main/java/org/apache/helix/controller/rebalancer/waged/model/AssignableNode.java

```diff
@@ -19,10 +19,293 @@
+          .noneMatch(type -> type.equals(faultZoneType))) {
+        throw new HelixException(
+            "The configured topology definition is empty or does not contain the fault zone type.");
+      }
+
+      Map<String, String> domainAsMap = instanceConfig.getDomainAsMap();
+      if (domainAsMap == null) {
+        throw new HelixException(String
+            .format("The domain configuration of instance %s is not configured", _instanceName));
+      } else {
+        StringBuilder faultZoneStringBuilder = new StringBuilder();
+        for (String key : topologyDef) {
+          if (!key.isEmpty()) {
+            if (domainAsMap.containsKey(key)) {
+              faultZoneStringBuilder.append(domainAsMap.get(key));
+              faultZoneStringBuilder.append('/');
+            } else {
+              throw new HelixException(String.format(
+                  "The domain configuration of instance %s is not complete. Type %s is not found.",
+                  _instanceName, key));
+            }
+            if (key.equals(faultZoneType)) {
+              break;
+            }
+          }
+        }
+        return faultZoneStringBuilder.toString();
+      }
+    } else {
+      // For backward compatibility
+      String zoneId = instanceConfig.getZoneId();
+      return zoneId == null ? instanceConfig.getInstanceName() : zoneId;
+    }
+  }
+
+  /**
+   * This function should only be used to assign a set of partitions that doesn't exist before.
+   * Using this function avoids the overhead of updating capacity repeatedly.
+   */
+  private void assignNewBatch(Collection<AssignableReplica> replicas) {
+    Map<String, Integer> totalPartitionCapacity = new HashMap<>();
+    for (AssignableReplica replica : replicas) {
+      addToAssignmentRecord(replica, _currentAssignments);
+      if (replica.isReplicaTopState()) {
+        addToAssignmentRecord(replica, _currentTopStateAssignments);
+      }
+      // increment the capacity requirement according to partition's capacity configure.
+      for (Map.Entry<String, Integer> capacity : replica.getCapacity().entrySet()) {
+        totalPartitionCapacity.compute(capacity.getKey(),
+            (k, v) -> (v == null) ? capacity.getValue() : v + capacity.getValue());
+      }
+    }
+    _totalReplicaAssignmentCount += replicas.size();
+
+    // Update to the global state after all single replications' calculation is done.
+    for (String key : totalPartitionCapacity.keySet()) {
+      updateCapacityAndUtilization(key, totalPartitionCapacity.get(key));
+    }
+  }
+
+  private boolean addToAssignmentRecord(AssignableReplica replica,
+      Map<String, Set<String>> currentAssignments) {
+    return currentAssignments.computeIfAbsent(replica.getResourceName(), k -> new HashSet<>())
+        .add(replica.getPartitionName());
+  }
+
+  private void updateCapacityAndUtilization(String capacityKey, int valueToSubtract) {
+    if (_currentCapacity.containsKey(capacityKey)) {
```

Comment:
What is the else condition? What if the given capacityKey doesn't exist? How do we alert?

Reply:
That basically means this category is unlimited on this node. This shall be allowed since node capacity and partition weight might be configured in a different logic. There is no need to require strictly consistent.

If adding any warning log here, the result will be very verbose.

The only restrict now is that node capacity categories have to match the cluster level definition.

Reply:
In that case, could we add this information as a block comment? 

`If the capacityKey does not exist in the capacity map, this method essentially becomes a NOP; in other words, this node will be treated as if it has unlimited capacity.`